crates/typst-utils/src/format.rs

```diff
@@ -0,0 +1,231 @@
+    (value * offset).round() / offset
+}
+
+/// Returns `(days, hours, minutes, seconds, milliseconds, microseconds)`.
+fn get_duration_parts(duration: &Duration) -> (u16, u8, u8, u8, u16, u16) {
+    // In practice we probably don't need nanoseconds.
+    let micros = duration.as_micros();
+    let (millis, micros) = (micros / 1000, (micros % 1000) as u16);
+    let (sec, millis) = (millis / 1000, (millis % 1000) as u16);
+    let (mins, sec) = (sec / 60, (sec % 60) as u8);
+    let (hours, mins) = (mins / 60, (mins % 60) as u8);
+    let (days, hours) = ((hours / 24) as u16, (hours % 24) as u8);
+    (days, hours, mins, sec, millis, micros)
+}
+
+/// Format string using `days`, `hours`, `minutes`, `seconds`.
+fn format_dhms(days: u16, hours: u8, minutes: u8, seconds: u8) -> String {
+    match (days, hours, minutes, seconds) {
+        (0, 0, 0, s) => format!("{s:2} s"),
+        (0, 0, m, s) => format!("{m:2} m {s:2} s"),
+        (0, h, m, s) => format!("{h:2} h {m:2} m {s:2} s"),
+        (d, h, m, s) => format!("{d:3} d {h:2} h {m:2} m {s:2} s"),
+    }
+}
+
+/// Format string starting with number of seconds and going bigger from there.
+///
+/// ```
+/// # use std::time::Duration;
+/// # use typst_utils::format::time_starting_with_seconds;
+/// let duration1 = time_starting_with_seconds(&Duration::from_secs(0));
+/// assert_eq!(" 0 s", &duration1);
+///
+/// let duration2 = time_starting_with_seconds(&Duration::from_secs(
+///     24 * 60 * 60 * 100 + // days
+///     60 * 60 * 10 + // hours
+///     60 * 10 + // minutes
+///     10 // seconds
+/// ));
+/// assert_eq!("100 d 10 h 10 m 10 s", &duration2);
+/// ```
+pub fn time_starting_with_seconds(duration: &Duration) -> String {
+    let (days, hours, minutes, seconds, _, _) = get_duration_parts(duration);
+    format_dhms(days, hours, minutes, seconds)
+}
+
+/// Format string starting with number of milliseconds and going bigger
+/// from there. `precision` is how many digits of microseconds
+/// from floating point to the right will be preserved (with rounding).
+/// Note that this function will always remove all trailing zeros
+/// for microseconds.
+///
+/// ```
+/// # use std::time::Duration;
+/// # use typst_utils::format::time_starting_with_milliseconds;
+/// let duration1 = time_starting_with_milliseconds(&Duration::from_micros(
+///     123 * 1000 + // milliseconds
+///     456 // microseconds
+/// ), 2);
+/// assert_eq!("123.46 ms", &duration1);
+///
+/// let duration2 = time_starting_with_milliseconds(&Duration::from_micros(
+///     123 * 1000 // milliseconds
+/// ), 2);
+/// assert_eq!("123 ms", &duration2);
+///
+/// let duration3 = time_starting_with_milliseconds(&Duration::from_secs(1), 2);
+/// assert_eq!(" 1 s", &duration3);
+/// ```
+///
+/// Note: if duration is 1 second or longer, then output will be identical
+/// to [time_starting_with_seconds], which also means that precision,
+/// number of milliseconds and microseconds will not be used.
```

Comment:
Let's move this to a separate paragraph before the example

Reply:
Yeah, but then there will be 2 _notes_.